File: 003.clj
(defn square [x] (* x x))
(defn defactor [n]
  (defn hlp [n divisor factor]
    (cond (> (square divisor) n) (cons n factor)
          (= 0 (rem n divisor)) (hlp (/ n divisor) divisor (cons divisor factor))
          :else (hlp n (+ divisor 1) factor)))
  (hlp n 2 '()))

(println (defactor 13195))
(println (defactor 600851475143))

File: 003.hs
defactor n = hlp n 2
    where hlp n i
              | i * i > n = [n]
              | n `mod` i == 0 = i: hlp (n `quot` i) i
              | otherwise = hlp n (i + 1)

main = do
        print $ defactor 13195
        print $ defactor 600851475143
